### PR TITLE
Adding Cascade Infer to FlashInfer

### PR DESCRIPTION
**Pull request by raywanb**
This draft pr is a work in progress aiming to add [cascade inference ](https://flashinfer.ai/2024/02/02/cascade-inference.html)to vllm. This is supposed to speedup inference when there are multiple requests that share the same prefix with a cold cache. Currently, this is still not working entirely: the output differs when not using cascade inference.  

**BEFORE SUBMITTING, PLEASE READ THE CHECKLIST BELOW AND FILL IN THE DESCRIPTION ABOVE**

---

<details>

<summary><b> PR Checklist (Click to Expand) </b></summary>

<p>Thank you for your contribution to vLLM! Before submitting the pull request, please ensure the PR meets the following criteria. This helps vLLM maintain the code quality and improve the efficiency of the review process.</p>

<h3>PR Title and Classification</h3>
<p>Only specific types of PRs will be reviewed. The PR title is prefixed appropriately to indicate the type of change. Please use one of the following:</p>
<ul>
    <li><code>[Bugfix]</code> for bug fixes.</li>
    <li><code>[CI/Build]</code> for build or continuous integration improvements.</li>
    <li><code>[Doc]</code> for documentation fixes and improvements.</li>
    <li><code>[Model]</code> for adding a new model or improving an existing model. Model name should appear in the title.</li>
    <li><code>[Frontend]</code> For changes on the vLLM frontend (e.g., OpenAI API server, <code>LLM</code> class, etc.) </li>
    <li><code>[Kernel]</code> for changes affecting CUDA kernels or other compute kernels.</li>
    <li><code>[Core]</code> for changes in the core vLLM logic (e.g., <code>LLMEngine</code>, <code>AsyncLLMEngine</code>, <code>Scheduler</code>, etc.)</li>
    <li><code>[Hardware][Vendor]</code> for hardware-specific changes. Vendor name should appear in the prefix (e.g., <code>[Hardware][AMD]</code>).</li>
    <li><code>[Misc]</code> for PRs that do not fit the above categories. Please use this sparingly.</li>
</ul>
<p><strong>Note:</strong> If the PR spans more than one category, please include all relevant prefixes.</p>

<h3>Code Quality</h3>

<p>The PR need to meet the following code quality standards:</p>

<ul>
    <li>We adhere to <a href="https://google.github.io/styleguide/pyguide.html">Google Python style guide</a> and <a href="https://google.github.io/styleguide/cppguide.html">Google C++ style guide</a>.</li>
    <li>Pass all linter checks. Please use <a href="https://github.com/vllm-project/vllm/blob/main/format.sh"><code>format.sh</code></a> to format your code.</li>
    <li>The code need to be well-documented to ensure future contributors can easily understand the code.</li>
    <li>Include sufficient tests to ensure the project to stay correct and robust. This includes both unit tests and integration tests.</li>
    <li>Please add documentation to <code>docs/source/</code> if the PR modifies the user-facing behaviors of vLLM. It helps vLLM user understand and utilize the new features or changes.</li>
</ul>

<h3>Notes for Large Changes</h3>
<p>Please keep the changes as concise as possible. For major architectural changes (>500 LOC excluding kernel/data/config/test), we would expect a GitHub issue (RFC) discussing the technical design and justification. Otherwise, we will tag it with <code>rfc-required</code> and might not go through the PR.</p>

<h3>What to Expect for the Reviews</h3>

<p>The goal of the vLLM team is to be a <i>transparent reviewing machine</i>. We would like to make the review process transparent and efficient and make sure no contributor feel confused or frustrated. However, the vLLM team is small, so we need to prioritize some PRs over others. Here is what you can expect from the review process: </p>

<ul>
    <li> After the PR is submitted, the PR will be assigned to a reviewer. Every reviewer will pick up the PRs based on their expertise and availability.</li>
    <li> After the PR is assigned, the reviewer will provide status update every 2-3 days. If the PR is not reviewed within 7 days, please feel free to ping the reviewer or the vLLM team.</li>
    <li> After the review, the reviewer will put an <code> action-required</code> label on the PR if there are changes required. The contributor should address the comments and ping the reviewer to re-review the PR.</li>
    <li> Please respond to all comments within a reasonable time frame. If a comment isn't clear or you disagree with a suggestion, feel free to ask for clarification or discuss the suggestion.
 </li>
</ul>

<h3>Thank You</h3>

<p> Finally, thank you for taking the time to read these guidelines and for your interest in contributing to vLLM. Your contributions make vLLM a great tool for everyone! </p>

</details>

### CODE REVIEW

**Comment by pavanimajety** [comment]
I am interested in looking at the perf improvement in attention layer compute before and after this change especially since we are removing the dependence on individual Prefill and Decode Wrappers. 

I suggest keeping them separate to allow us to (1) use different kernels if required in future based on model or use-case and (2) to continue to have support for cuda graphs.

**Comment by comaniac** [comment]
> I suggest keeping them separate to allow us to (1) use different kernels if required in future based on model or use-case and (2) to continue to have support for cuda graphs. 
> 
 We intentionally want to use a single kernel for all to simply the implementation. Meanwhile, we of course want to make sure there won't be performance regression. For the CUDA graph support, @yzh119 is planning to support it in FlashInfer in these 2 days and will make a new release, so we ideally will have CUDA graph support before merging this PR.

File: vllm/attention/backends/flashinfer.py [hunk 1]
```diff
@@ -552,26 +537,73 @@ def _add_seq_group(
...
             last_page_len = self.block_size
-        self.paged_kv_last_page_len.append(last_page_len)
+        self.second_layer_kv_last_page_len.append(last_page_len)
+
+    def _update_shared_kv_tensors(self, common_prefix: List[int],
+                                  batch_size: int) -> None:
+        """
+        Updates the shared level kv tensors
+        """
+        if len(common_prefix) == 0:
```
Comment (comaniac) [inline vllm/attention/backends/flashinfer.py:568]:
```suggestion
        if not common_prefix:
```

File: vllm/attention/backends/flashinfer.py [hunk 2]
```diff
@@ -552,26 +537,73 @@ def _add_seq_group(
...
+        Returns a list of consecutive shared blocks across sequence groups
+        """
+        block_id_lists = [
+            list(data.block_tables.values()) for data in inter_data_list
+            if data.block_tables
+        ]
+
+        flattened_lists = [
+            item for sublist in block_id_lists for item in sublist
+        ]
```
Comment (comaniac) [inline vllm/attention/backends/flashinfer.py:597]: Should be more efficient:
```suggestion
        flattened_lists = []
        for data in inter_data_list:
            if data.block_tables:
                flattened_lists += data.block_tables.values()
```

File: vllm/attention/backends/flashinfer.py [hunk 3]
```diff
@@ -584,9 +616,19 @@ def build(self, seq_lens: List[int], query_lens: List[int],
                                  -1 if cuda graph is not used.
             batch_size: The maybe padded batch size.
         """
+        # common_prefix = self.get_shared_blocks_nums(
+        #     self.input_builder.inter_data_list
+        # )
+        # currently, we set common_prefix to empty list since
+        # shared level is not working yet.
```
Comment (comaniac) [inline vllm/attention/backends/flashinfer.py:623]:
```suggestion
        # FIXME: we set common_prefix to empty list now since
        # shared level is not working yet.
```

File: vllm/attention/backends/flashinfer.py [hunk 4]
```diff
@@ -645,20 +687,41 @@ def build(self, seq_lens: List[int], query_lens: List[int],
...
                      dtype=seq_start_loc.dtype,
                      out=seq_start_loc[1:])
+
         torch.cumsum(query_lens_tensor,
                      dim=0,
                      dtype=query_start_loc.dtype,
-                     out=query_start_loc[1:])
+                     out=second_level_query_start_loc[1:])
+
+        if len(common_prefix) == 0:
```
Comment (comaniac) [inline vllm/attention/backends/flashinfer.py:705]:
```suggestion
        if not common_prefix:
```

File: vllm/attention/backends/flashinfer.py [hunk 5]
```diff
@@ -770,6 +850,9 @@ def forward(
         if attn_metadata.num_decode_tokens > 0:
             assert attn_metadata.num_prefill_tokens == 0, (
                 "Chunked prefill is not supported with flashinfer yet.")
+
+        attn_metadata.plan(self.scale, self.logits_soft_cap)
```
Comment (comaniac) [inline vllm/attention/backends/flashinfer.py:854]: Per offline discussion, should we just call this once?

File: vllm/attention/backends/abstract.py [hunk 6]
```diff
@@ -181,7 +182,8 @@ def prepare_graph_input_buffers(
         ...
 
     @abstractmethod
-    def begin_forward(self, model_input: "ModelRunnerInputBase") -> None:
+    def plan(self, model_input: "ModelRunnerInputBase",
```
Comment (comaniac) [inline vllm/attention/backends/abstract.py:185]: You will need to change other attention implementation's API as well. Alternative you can keep the original name. It's up to you.

**Review by comaniac — COMMENTED** [review]
Overall LGTM. Waiting for CUDA graph support from @yzh119
Meanwhile, could you run end to end serving and throughput benchmarks on H100 to make sure there's no performance regression (without CUDA graph for now)?
